2.MLP_Multi_Classifier/MLP_utils/utils.py

```diff
@@ -0,0 +1,938 @@
+
+    # optimizer = (
+    # f'optim.{optim_method}(model.parameters(), lr=parameter_dict["learning_rate"])'
+    # )
+    # print(optimizer)
+    optimizer = eval(optimizer)
+    # optimizer = ast.literal_eval(optimizer)
+    print(optimizer)
+
+    early_stopping_patience = 15
+    early_stopping_counter = 0
+
+    train_acc = []
+    train_loss = []
+
+    valid_acc = []
+    valid_loss = []
+
+    total_step = len(train_loader)
+    total_step_val = len(valid_loader)
+
+    valid_loss_min = np.inf
+
+    epochs_ran = []
+
+    for epoch in range(EPOCHS):
+
+        epochs_ran.append(epoch + 1)
+
+        (
+            model,
+            optimizer,
+            criterion,
+            train_acc,
+            train_loss,
+            valid_acc,
+            valid_loss,
+            correct,
+            total,
+            correct_v,
+            total_v,
+        ) = train_n_validate(
+            model,
+            optimizer,
+            criterion,
+            train_acc,
+            train_loss,
+            valid_acc,
+            valid_loss,
+            total_step,
+            total_step_val,
+            params,
+            train_loader,
+            valid_loader,
+        )
+
+        if np.mean(valid_loss) <= valid_loss_min:
+            torch.save(model.state_dict(), f"./state_dict.pt")
+            print(
+                f"Epoch {epoch + 0:01}: Validation loss decreased ({valid_loss_min:.6f} --> {np.mean(valid_loss):.6f}).  Saving model ..."
+            )
+            valid_loss_min = np.mean(valid_loss)
+            early_stopping_counter = 0  # reset counter if validation loss decreases
+        else:
+            print(f"Epoch {epoch + 0:01}: Validation loss did not decrease")
+            early_stopping_counter += 1
+
+        if early_stopping_counter > early_stopping_patience:
+            print("Early stopped at epoch :", epoch)
+            break
+
+        print(
+            f"\t Train_Loss: {np.mean(train_loss):.4f} Train_Acc: {(100 * correct / total):.3f} Val_Loss: {np.mean(valid_loss):.4f}  BEST VAL Loss: {valid_loss_min:.4f}  Val_Acc: {(100 * correct_v / total_v):.3f}\n"
+        )
+    return train_loss, train_acc, valid_loss, valid_acc, epochs_ran, model
+
+
+# Plot training data over epochs
+def plot_metric_vs_epoch(df: pd.DataFrame, x: str, y1: str, y2: str) -> None:
+    """Plot x vs y1 and x vs y2 using seaborn.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        pandas data frame of choice
+    x : str
+        x variable in this case epoch number
+    y1 : str
+        y1 variable in this case train metric
+    y2 : str
+        y1 variable in this case validation metric
+    """
+    sns.lineplot(x=x, y=y1, data=df)
+    sns.lineplot(x=x, y=y2, data=df)
+
+
+def test_optimized_model(
+    model: torch.nn.Sequential,
+    test_loader: torch.utils.data.DataLoader,
+    in_features: int,
+    out_features: int,
+    parameter_dict: dict,
+    params: object,
```

Comment:
Consider renaming this parameter for greater specificity and possibly referencing the the related dataclass for the type hint (instead of a generic object). This comment applies to several functions in this file (see areas where `params: object` is used).

Reply:
addressed and changed